### ISSUE TITLE
Unable to use %{env:FTPS} in a SQLNamedQuery

### ISSUE THREAD

**Issue by conathan** [proftpd/proftpd#1232]
We have the following in proftpd.conf, which was working as expected on Centos 7/EPEL (proftpd 1.3.5e), but when we moved to AlmaLinux 8/EPEL (proftpd 1.3.6e) it stopped working (%{env:FTPS} kept being replaced by '')

`SQLNamedQuery get_mod_sql_userinfo FREEFORM "SELECT username, password, unix_id as uid, unix_id as gid, "/home/test" as homedir, null as shell FROM ftp_accounts left join users ON ftp_accounts.user_id = users.id WHERE username='%U' AND ftp_accounts.status = 2 AND (legacy_ftp = 1 OR '%{env:FTPS}' = '1')"`

Idea behind this, is that we can enable/disable the ability for specific users to login via unencrypted FTP, while still allowing them to login via FTPS.  (mod_tls sets FTPS to 1, so we rely on that for the statement to work)

In proftpd 1.3.5e, %{env:FTPS} would result in '1' for users using ftps, '' for unencrypted ftp users,
while 1.3.6e %{env:FTPS} is '' for both ftp and ftps.
* Going forward, will use '' in place of a blank string

-----
In diagnosing this, It feels like something is processing and replacing %{env:FTPS} in the config file, before this is handed over to mod_sql (While in our usecase, it would only work if this is done when the query is processed, as FTPS would be set per connection).

In contrib/mod_sql.c, static modret_t *process_named_query(cmd_rec *cmd, char *name, int flags)

I added the following:

```
   /* Check for a query by that name. */
   query = pstrcat(cmd->tmp_pool, "SQLNamedQuery_", name, NULL);
 
   c = find_config(main_server->conf, CONF_PARAM, query, FALSE);
+  sql_log(DEBUG_FUNC, "!!!2 '%s'", c->argv[1]);
   if (c != NULL) {
     size_t arglen, outs_remain = sizeof(outs)-1;
```

And in the sql log file I see:
`"!!!2 SELECT username, password, unix_id as uid, unix_id as gid, "/home/test" as homedir, null as shell FROM ftp_accounts left join users ON ftp_accounts.user_id = users.id WHERE username='%U' AND ftp_accounts.status = 2 AND (legacy_ftp = 1 OR '' = '1')"`

(basically, %{env:FTPS} is missing).  This is before it runs through the string, looking for %{
*  before   for (tmp = c->argv[1]; *tmp; ) {

----

To test this further, I also made the following change in mod_sql.c to allow one to use either %{env:FTPS} or &{env:FTPS}

```
    for (tmp = c->argv[1]; *tmp; ) {
      char *tag = NULL;

-      if (*tmp == '%') {
+      if (*tmp == '%' || *tmp == '&') {
        if (*(++tmp) == '{') {
          char *tmp_query = NULL;

          if (*tmp != '\0') {
            tmp_query = ++tmp;
          }
```

And now if I rewrite the SQLNamedQuery above to: (using &{env:FTPS}
`SQLNamedQuery get_mod_sql_userinfo FREEFORM "SELECT username, password, unix_id as uid, unix_id as gid, "/home/test" as homedir, null as shell FROM ftp_accounts left join users ON ftp_accounts.user_id = users.id WHERE username='%U' AND ftp_accounts.status = 2 AND (legacy_ftp = 1 OR '&{env:FTPS}' = '1')"`

It works as expected again.

In conclusion, %{env:FTPS} is not working in a SQLNamedQuery, as it used to in 1.3.5e.

----
Note: Change is not meant to be a proposed patch, but just as a tool to illustrate what I've done to debug this so far.

**Comment by Castaglia** [comment]
I believe this may be related to Issue #857; are you able to use a newer ProFTPD version?

**Comment by Castaglia** [comment]
I'm able to reproduce this behavior locally now; it is indeed looking like the config file parser is "eating" that `%{env:FTPS}` string.  I think I've seen this behavior before...

**Comment by Castaglia** [comment]
Once https://github.com/proftpd/proftpd/pull/1234 is merged, the fix will unfortunately require a change to your existing configuration.  Rather than using `%{env:FTPS}` in your `SQLNamedQuery`, you should be able to use `%{note:mod_tls.FTPS}` to restore the previous behavior.

**Comment by Castaglia** [comment]
Actually, I'm going to tweak this fix a little, so that the note variables use the same names as the environment variables; this is how it is already done in _e.g._ `mod_geoip`.  That is, once I've updated the fix, you use `%{note:FTPS}` in your `SQLNamedQuery` instead of `%{env:FTPS}`.

**Comment by conathan** [comment]
That looks like it will work perfectly for us, Thank you (And no issues on our side changing config file to make this work).

As for proftpd 1.3.6 vs 1.3.7, This is on AlmaLinux 8 using EPEL repositories.  They are using 1.3.6e at this time.  (Well to be fair we have our own custom patches and recompile the srpm, but I am trying to keep pace with what they have until I can get rid of our custom plugins).

**Comment by Castaglia** [comment]
Thanks for the confirmation!  I'll now close this.
